subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskExecutionStatisticsEventAdapter.java

```diff
@@ -16,34 +16,28 @@
 
 package org.gradle.caching.internal.tasks;
 
-import com.google.common.base.Functions;
-import com.google.common.collect.Maps;
 import org.gradle.api.Task;
 import org.gradle.api.execution.TaskExecutionListener;
-import org.gradle.api.internal.tasks.TaskExecutionOutcome;
-import org.gradle.api.internal.tasks.TaskStateInternal;
 import org.gradle.api.tasks.TaskState;
-import org.gradle.caching.internal.tasks.statistics.TaskExecutionStatistics;
-import org.gradle.caching.internal.tasks.statistics.TaskExecutionStatisticsListener;
 import org.gradle.initialization.BuildCompletionListener;
-
-import java.util.Arrays;
-import java.util.Map;
+import org.gradle.internal.statistics.TaskExecutionStatistics;
+import org.gradle.internal.statistics.TaskExecutionStatisticsAggregator;
+import org.gradle.internal.statistics.TaskExecutionStatisticsReporter;
 
 public class TaskExecutionStatisticsEventAdapter implements BuildCompletionListener, TaskExecutionListener {
-    private final TaskExecutionStatisticsListener listener;
-    private final Map<TaskExecutionOutcome, Integer> taskCounts = Maps.newEnumMap(
-        Maps.toMap(Arrays.asList(TaskExecutionOutcome.values()), Functions.constant(0))
-    );
-    private int cacheMissCount;
+    private final TaskExecutionStatisticsAggregator statisticsAggregator;
+    private final TaskExecutionStatisticsReporter listener;
 
-    public TaskExecutionStatisticsEventAdapter(TaskExecutionStatisticsListener listener) {
+    public TaskExecutionStatisticsEventAdapter(TaskExecutionStatisticsReporter listener) {
         this.listener = listener;
+        this.statisticsAggregator = new TaskExecutionStatisticsAggregator();
     }
 
     @Override
     public void completed() {
-        listener.buildFinished(new TaskExecutionStatistics(taskCounts, cacheMissCount));
+        final TaskExecutionStatistics stats = new TaskExecutionStatistics(
+            statisticsAggregator.getTaskCounts(), statisticsAggregator.getCacheMissCount());
```

Comment:
Could we (and should we) avoid creating a new `TaskExecutionStatistics` by having a method on `TaskExecutionStatisticsAggregator` to return an `TaskExecutionStatistics`. It would avoid creating a temporary object just to transfer the statistic and, instead, have a permanent object inside `TaskExecutionStatisticsAggregator`? Something along the line of `toStatistics()`.

Here we only create the `TaskExecutionStatistics` once the build is completed but for the progress bar, in the worth case, we will create one every 85ms. What do you think?